File: Data Structures & Algorithms/add-two-numbers/submission-1.rs
impl Solution {
    pub fn add_two_numbers(
        mut l1: Option<Box<ListNode>>,
        mut l2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        let mut dummy = Box::new(ListNode::new(0));
        let mut tail = &mut dummy;
        let mut carry = 0;

        while l1.is_some() || l2.is_some() || carry != 0 {
            let mut total = carry;
            if let Some(mut node) = l1 {
                total += node.val;
                l1 = node.next.take();
            }
            if let Some(mut node) = l2 {
                total += node.val;
                l2 = node.next.take();
            }

            carry = total / 10;
            tail.next = Some(Box::new(ListNode::new(total % 10)));
            tail = tail.next.as_mut().unwrap();
        }

        dummy.next
    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.py
import heapq


class Solution:
    def mergeKLists(self, lists):
        heap = []

        for index, node in enumerate(lists):
            if node is not None:
                heapq.heappush(heap, (node.val, index, node))

        dummy = ListNode(0)
        tail = dummy

        while heap:
            _, index, node = heapq.heappop(heap)
            tail.next = node
            tail = node

            if node.next is not None:
                heapq.heappush(heap, (node.next.val, index, node.next))

        return dummy.next